### PR TITLE
Add ANN optimize, Clean code and Fix some bugs

### PR DESCRIPTION
**Pull request by nghi-truyen**
- Merge net.compile and net.fit to simplify the use of model.optimize with ann mapping
- Group certain params in optimize_options into a dict "termination_crit"
- Separate py_standardize and finalize_standardize. The first one is only served for immediately standardize while the second is for Fortran standardization.
- Handle with options constants (None and default values) 
- New functions in options.py which will be served for verify Py standardization and documentation of model.optimize, model.run, etc.
- Add new test for ann optimize and regenerate baseline
- Taking into account PR #55 for Issue #53 

TODO:
- Handle with temporal solution for dealloc Fortran derived type
- Continue cleaning: constant and standardize functions

### CODE REVIEW

File: smash/factory/net/net.py [hunk 1]
```diff
@@ -2,22 +2,19 @@
 
 from smash._typing import Numeric
 
-# from smash.solver._mw_forward import forward_b
+from smash._constant import OPTIMIZER_CLASS, PY_OPTIMIZER
```
Comment (inoelloc) [inline smash/factory/net/net.py:5]:
```suggestion
from smash._constant import PY_OPTIMIZER_CLASS, PY_OPTIMIZER
```

File: smash/factory/net/net.py [hunk 2]
```diff
@@ -242,11 +228,48 @@ def add(self, layer: str, options: dict):
...
+                since they do not involve any weight updates during training.
+        """
+
+        if len(trainable) == len(self.layers):
+            for i, layer in enumerate(self.layers):
+                layer.trainable = trainable[i]
+
+        else:
+            raise ValueError(
+                f"Inconsistent length between trainable ({len(trainable)}) and the number of layers ({len(self.layers)})"
```
Comment (inoelloc) [inline smash/factory/net/net.py:265]: ? comme tu veux 
```suggestion
                f"Inconsistent size between trainable ({len(trainable)}) and the number of layers ({len(self.layers)})"
```

File: smash/factory/net/net.py [hunk 3]
```diff
@@ -304,11 +327,11 @@ def compile(
         """
 
         if self.layers:
-            optimizer, random_state, options = _standardize_compile_args(
-                optimizer, random_state, options
-            )
+            ind = PY_OPTIMIZER.index(optimizer.lower())
+
+            func = eval(OPTIMIZER_CLASS[ind])
```
Comment (inoelloc) [inline smash/factory/net/net.py:332]:
```suggestion
            func = eval(PY_OPTIMIZER_CLASS[ind])
```

File: smash/fcore/optimize/mw_control_dealloc.f90
Comment (inoelloc) [inline smash/fcore/optimize/mw_control_dealloc.f90:1]: A discuter

File: smash/core/simulation/optimize/optimize.py [hunk 4]
```diff
@@ -64,12 +75,77 @@ def _optimize(
...
+    model: Model,
+    optimizer: str,
+    optimize_options: dict,
+    common_options: dict,
+    wrap_options: OptionsDT,
+    wrap_returns: ReturnsDT,
+):
+    # % preprocessing input descriptors and normalization
+    active_mask = np.where(model.mesh.active_cell == 1)
+    inactive_mask = np.where(model.mesh.active_cell == 0)
```
[the remark this answers is not in the document]
Comment (nghi-truyen) [reply smash/core/simulation/optimize/optimize.py:110]: je crois que ça marche pas comme ça pour une tuple

**Review by inoelloc — APPROVED** [review]

File: smash/core/simulation/optimize/optimize.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (inoelloc) [reply smash/core/simulation/optimize/optimize.py:110]: Yes je pensais que np.where renvoyait un array. On laisse comme ca ;)

**Pull request merged**